controllers/infrastructure/byocluster_controller_test.go

```diff
@@ -0,0 +1,135 @@
+		k8sClientUncached client.Client
+		byoCluster        *infrastructurev1beta1.ByoCluster
+		cluster           *clusterv1.Cluster
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		var clientErr error
+
+		k8sClientUncached, clientErr = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+		Expect(clientErr).NotTo(HaveOccurred())
+	})
+
+	It("should ignore when byocluster is not existed", func() {
+		_, err := byoClusterReconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      "non-existent-byocluster",
+				Namespace: "non-existent-namespace"}})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should ignore When Cluster Controller is not set OwnerRef on ByoCluster", func() {
+		byoCluster = builder.ByoCluster(defaultNamespace, "byocluster-not-link-cluster").Build()
+		Expect(k8sClientUncached.Create(ctx, byoCluster)).Should(Succeed())
+		WaitForObjectsToBePopulatedInCache(byoCluster)
+
+		_, err := byoClusterReconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      byoCluster.Name,
+				Namespace: byoCluster.Namespace}})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should paused when byocluster is paused", func() {
+		cluster = builder.Cluster(defaultNamespace, "cluster-paused").
+			WithPausedField(true).
+			Build()
+		Expect(k8sClientUncached.Create(ctx, cluster)).Should(Succeed())
+		WaitForObjectsToBePopulatedInCache(cluster)
+
+		byoCluster = builder.ByoCluster(defaultNamespace, "byocluster-paused").
+			WithOwnerCluster(cluster).
+			Build()
+		Expect(k8sClientUncached.Create(ctx, byoCluster)).Should(Succeed())
+
+		_, err := byoClusterReconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      byoCluster.Name,
+				Namespace: byoCluster.Namespace}})
+		Expect(err).NotTo(HaveOccurred())
+
+	})
+
+	It("should add ClusterFinalizer on ByoCluster", func() {
+		cluster = builder.Cluster(defaultNamespace, "byocluster-finalizer").
+			Build()
+		Expect(k8sClientUncached.Create(ctx, cluster)).Should(Succeed())
+		WaitForObjectsToBePopulatedInCache(cluster)
+
+		byoCluster = builder.ByoCluster(defaultNamespace, "byocluster-finalizer").
+			WithOwnerCluster(cluster).
+			Build()
+		Expect(k8sClientUncached.Create(ctx, byoCluster)).Should(Succeed())
+		WaitForObjectsToBePopulatedInCache(byoCluster)
+
+		byoClusterLookupKey := types.NamespacedName{Name: byoCluster.Name, Namespace: byoCluster.Namespace}
+		_, err := byoClusterReconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: byoClusterLookupKey})
+		Expect(err).NotTo(HaveOccurred())
+
+		updatedByoCluster := &infrastructurev1beta1.ByoCluster{}
+		err = k8sClientUncached.Get(ctx, byoClusterLookupKey, updatedByoCluster)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(controllerutil.ContainsFinalizer(updatedByoCluster, infrastructurev1beta1.ClusterFinalizer)).To(BeTrue())
+
+	})
+
+	It("should be able to delete ByoCluster", func() {
```

Comment:
This can move under  a separate Context
```
Context("when ByoClusterController gets a delete request", func(){

    It("should wait for byomachines to be deleted", func() {

    }

    It("should be able to delete ByoCluster", func() {

    }
})
```